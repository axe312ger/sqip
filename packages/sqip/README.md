# `sqip`

> TODO: description

## Usage

```
import sqip from 'sqip';

// TODO: DEMONSTRATE API
```
