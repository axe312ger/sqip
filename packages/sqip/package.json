{
  "name": "sqip",
  "version": "1.0.0-alpha.32",
  "description": "> TODO: description",
  "author": "Benedikt RÃ¶tsch <info@benedikt-roetsch.de>",
  "homepage": "https://github.com/axe312ger/sqip#readme",
  "license": "MIT",
  "scripts": {
    "build": "tsc -b",
    "watch": "tsc -w",
    "test": "echo \"Error: run tests from root\" && exit 1"
  },
  "engines": {
    "node": ">=12.13.0"
  },
  "types": "dist/sqip.d.ts",
  "main": "dist/sqip.js",
  "directories": {
    "lib": "dist",
    "test": "__tests__"
  },
  "files": [
    "dist"
  ],
  "repository": {
    "type": "git",
    "url": "git+https://github.com/axe312ger/sqip.git"
  },
  "bugs": {
    "url": "https://github.com/axe312ger/sqip/issues"
  },
  "keywords": [
    "lqip",
    "svg",
    "images",
    "placeholder",
    "webperf",
    "performance",
    "ux",
    "perceived"
  ],
  "dependencies": {
    "chalk": "^4.1.0",
    "cheerio": "^1.0.0-rc.3",
    "cli-table3": "^0.6.0",
    "expand-tilde": "^2.0.2",
    "fast-glob": "^3.2.4",
    "fs-extra": "^9.0.1",
    "node-vibrant": "^3.2.1-alpha.1",
    "probe-image-size": "^7.1.0",
    "sharp": "^0.28.1",
    "term-img": "^5.0.0"
  },
  "peerDependencies": {
    "sqip-plugin-blur": "^1.0.0-alpha.0",
    "sqip-plugin-data-uri": "^1.0.0-alpha.0",
    "sqip-plugin-primitive": "^1.0.0-alpha.0",
    "sqip-plugin-svgo": "^1.0.0-alpha.0"
  },
  "devDependencies": {
    "@types/debug": "^4.1.5",
    "@types/expand-tilde": "^2.0.0",
    "@types/fs-extra": "^9.0.11",
    "@types/probe-image-size": "^7.0.0",
    "debug": "^4.1.1"
  }
}
