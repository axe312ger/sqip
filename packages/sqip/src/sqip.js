// #############################################################################
// #
// # "SQIP" (pronounced \skwÉªb\ like the non-magical folk of magical descent)
// # is a SVG-based LQIP technique - https://github.com/technopagan/sqip
// #
// # Installation as CLI:
// # npm install -g sqip
// #
// # Installation as lib:
// # npm install sqip
// #
// # Requirements:
// # * Node.js >= v6 (https://nodejs.org/en/)
// #
// #############################################################################

import path from 'path'

import Debug from 'debug'
import fs from 'fs-extra'

import { getDimensions, printFinalResult } from './helpers'

const debug = Debug('sqip')

export default async function sqip(options) {
  // Build configuration based on passed options and default options
  const defaultOptions = {
    plugins: [
      { name: 'primitive', options: { numberOfPrimitives: 8, mode: 0 } },
      'blur',
      'svgo',
      'data-uri'
    ],
    shouldThrow: true // @todo do we really need this?
  }
  const config = Object.assign({}, defaultOptions, options)

  // Validate configuration
  if (!config.input) {
    throw new Error(
      'Please provide an input image, e.g. sqip({ input: "input.jpg" })'
    )
  }

  const inputPath = path.resolve(process.cwd(), config.input)

  try {
    await fs.access(inputPath, fs.constants.R_OK)
  } catch (err) {
    throw new Error(`Unable to read input file: ${inputPath}`)
  }

  const dimensions = getDimensions(inputPath)

  // Resolver
  const plugins = await Promise.all(
    config.plugins.map(async plugin => {
      if (typeof plugin === 'string') {
        plugin = { name: plugin }
      }
      const { name } = plugin

      if (!name) {
        throw new Error(
          `Unable to read plugin name from:\n${JSON.stringify(plugin, null, 2)}`
        )
      }
      const moduleName =
        name.indexOf('sqip-plugin-') !== -1 ? name : `sqip-plugin-${name}`
      try {
        debug(`Loading ${moduleName}`)
        const Plugin = await import(moduleName)

        return { ...plugin, Plugin }
      } catch (err) {
        if (err.code === 'MODULE_NOT_FOUND') {
          throw new Error(
            `Unable to load plugin "${moduleName}". Try installing it via:\n\n npm install ${moduleName}`
          )
        }
        throw err
      }
    })
  )

  let svg = config.filename
  for (const { name, options: pluginOptions, Plugin } of plugins) {
    try {
      debug(`Construct ${name}`)
      const plugin = new Plugin({ dimensions, ...options, ...pluginOptions })
      debug(`Apply ${name}`)
      svg = await plugin.apply(svg, { dimensions })
    } catch (err) {
      if (config.shouldThrow) {
        throw err
      }
      console.error(err)
      process.exit(1)
    }
  }

  debug(`Finished`)

  // Write to disk or output result
  if (config.output) {
    const outputPath = path.resolve(config.output)
    // @todo remove all sync calls
    fs.writeFileSync(outputPath, svg)
  } else {
    printFinalResult(dimensions, inputPath, svg)
  }
  return { svg, dimensions }
}
