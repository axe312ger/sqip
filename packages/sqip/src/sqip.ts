import path from 'path'

import Debug from 'debug'
import fs from 'fs-extra'
import imageSize from 'probe-image-size'
import Vibrant from 'node-vibrant'
import sharp from 'sharp'
import termimg from 'term-img'
import Table from 'cli-table3'
import chalk from 'chalk'

import { Palette } from '@vibrant/color'
import { OptionDefinition } from 'command-line-args'

import { locateFiles } from './helpers'

export { loadSVG, parseColor } from './helpers'

const debug = Debug('sqip')

const mainKeys = ['originalWidth', 'originalHeight', 'width', 'height', 'type']

const PALETTE_KEYS: (keyof Palette)[] = [
  'Vibrant',
  'DarkVibrant',
  'LightVibrant',
  'Muted',
  'DarkMuted',
  'LightMuted'
]

interface SqipPluginInterface {
  metadata: SqipImageMetadata
  sqipConfig: SqipConfig
  apply(imageBuffer: Buffer): Promise<Buffer> | Buffer
}

export interface SqipResult {
  content: Buffer
  metadata: SqipImageMetadata
}

export interface SqipPluginOptions {
  pluginOptions: PluginOptions
  options: PluginOptions
  metadata: SqipImageMetadata
  sqipConfig: SqipConfig
}
export interface SqipCliOptionDefinition extends OptionDefinition {
  description?: string
  required?: boolean
  default?: boolean
}

export interface PluginOptions {
  [key: string]: unknown
}

interface PluginResolver {
  name: string
  options?: PluginOptions
}

interface SqipConfig {
  input: string | Buffer
  outputFileName?: string
  output: string
  silent?: boolean
  parseableOutput?: boolean
  plugins: PluginType[]
  width?: number
}

interface ProcessFileOptions {
  buffer: Buffer
  outputFileName: string
  config: SqipConfig
}

interface SqipImageMetadata {
  originalWidth: number
  originalHeight: number
  palette: Palette
  height: number
  width: number
  type: string
  [key: string]: unknown
}

type PluginType = PluginResolver | string

export class SqipPlugin implements SqipPluginInterface {
  public metadata: SqipImageMetadata
  public sqipConfig: SqipConfig
  public options: PluginOptions
  static cliOptions: SqipCliOptionDefinition[]

  constructor(options: SqipPluginOptions) {
    const { sqipConfig, metadata } = options
    this.sqipConfig = sqipConfig || {}
    this.metadata = metadata || {}
    this.options = {}
  }
  apply(imageBuffer: Buffer): Promise<Buffer> | Buffer {
    throw new Error('Not implemented')
    return imageBuffer
  }
}

// Resolves plugins based on a given config
// Array of plugin names or config objects, even mixed.
export async function resolvePlugins(
  plugins: PluginType[]
): Promise<(PluginResolver & { Plugin: typeof SqipPlugin })[]> {
  return Promise.all(
    plugins.map(async (plugin) => {
      if (typeof plugin === 'string') {
        plugin = { name: plugin }
      }
      const { name } = plugin

      if (!name) {
        throw new Error(
          `Unable to read plugin name from:\n${JSON.stringify(plugin, null, 2)}`
        )
      }
      const moduleName =
        name.indexOf('sqip-plugin-') !== -1 ? name : `sqip-plugin-${name}`
      try {
        debug(`Loading ${moduleName}`)
        const Plugin = await import(moduleName)

        return { ...plugin, Plugin: Plugin.default }
      } catch (err) {
        if (err.code === 'MODULE_NOT_FOUND') {
          throw new Error(
            `Unable to load plugin "${moduleName}". Try installing it via:\n\n npm install ${moduleName}`
          )
        }
        throw err
      }
    })
  )
}

async function processFile({
  buffer,
  outputFileName,
  config
}: ProcessFileOptions) {
  const { output, silent, parseableOutput } = config
  const result = await processImage({ buffer, config })
  const { content, metadata } = result
  let outputPath

  debug(`Processed ${outputFileName}`)

  // Write result svg if desired
  if (output) {
    try {
      // Test if output path already exists
      const stats = await fs.stat(output)

      // Throw if it is a file and already exists
      if (!stats.isDirectory()) {
        throw new Error(
          `File ${output} already exists. Overwriting is not yet supported.`
        )
      }
      outputPath = path.resolve(output, `${outputFileName}.svg`)
    } catch (err) {
      // Output directory or file does not exist. We will create it later on.
      outputPath = output
    }

    debug(`Writing ${outputPath}`)
    await fs.writeFile(outputPath, content)
  }

  // Gather CLI output information
  if (!silent) {
    if (outputPath) {
      console.log(`Stored at: ${outputPath}`)
    }

    // Generate preview
    if (!parseableOutput) {
      // Convert to png for image preview
      const preview = await sharp(Buffer.from(content)).png().toBuffer()

      try {
        termimg(preview, {
          fallback: () => {
            // SVG results can still be outputted as string
            if (metadata.type === 'svg') {
              console.log(content.toString())
              return
            }

            // No fallback preview solution yet for non-svg files.
            console.log(
              `Unable to render a preview for ${metadata.type} files on this machine. Try using https://iterm2.com/`
            )
          }
        })
      } catch (err) {
        if (err.name !== 'UnsupportedTerminalError') {
          throw err
        }
      }
    }

    // Metadata
    const tableConfig = parseableOutput
      ? {
          chars: {
            top: '',
            'top-mid': '',
            'top-left': '',
            'top-right': '',
            bottom: '',
            'bottom-mid': '',
            'bottom-left': '',
            'bottom-right': '',
            left: '',
            'left-mid': '',
            mid: '',
            'mid-mid': '',
            right: '',
            'right-mid': '',
            middle: ' '
          },
          style: { 'padding-left': 0, 'padding-right': 0 }
        }
      : undefined

    // Figure out which metadata keys to show
    // @todo why is this unused?
    // const allKeys = [...mainKeys, 'palette']

    const mainTable = new Table(tableConfig)
    mainTable.push(mainKeys)
    mainTable.push(
      mainKeys.map((key) => String(metadata[key]) || 'can not display')
    )
    console.log(mainTable.toString())

    // Show color palette
    const paletteTable = new Table(tableConfig)
    paletteTable.push(PALETTE_KEYS)
    paletteTable.push(
      PALETTE_KEYS.map((key) => metadata.palette[key]?.hex)
        .filter<string>((hex): hex is string => typeof hex === 'string')
        .map((hex) => chalk.hex(hex)(hex))
    )
    console.log(paletteTable.toString())

    Object.keys(metadata)
      .filter((key) => ![...mainKeys, 'palette'].includes(key))
      .forEach((key) => {
        console.log(chalk.bold(`${key}:`))
        console.log(metadata[key])
      })
  }

  return result
}

interface ProcessImageOptions {
  buffer: Buffer
  config: SqipConfig
}

async function processImage({
  buffer,
  config
}: ProcessImageOptions): Promise<SqipResult> {
  const originalSizes = imageSize.sync(buffer)
  const vibrant = Vibrant.from(buffer)
  const palette = await vibrant.quality(0).getPalette()

  if (!originalSizes) {
    throw new Error('Unable to get image size')
  }

  const metadata: SqipImageMetadata = {
    originalWidth: originalSizes.width,
    originalHeight: originalSizes.height,
    palette,
    type: 'unknown',
    width: 0,
    height: 0
  }

  // Load plugins
  const plugins = await resolvePlugins(config.plugins)

  // Interate through plugins and apply them to last returned image
  if (config.width && config.width > 0) {
    // Resize to desired output width
    buffer = await sharp(buffer).resize(config.width).toBuffer()

    const resizedMetadata = await sharp(buffer).metadata()
    metadata.width = resizedMetadata.width || 0
    metadata.height = resizedMetadata.height || 0
  } else {
    // Fall back to original size, keep image as is
    metadata.width = originalSizes.width
    metadata.height = originalSizes.height
  }

  for (const { name, options: pluginOptions, Plugin } of plugins) {
    debug(`Construct ${name}`)
    const plugin = new Plugin({
      sqipConfig: config,
      pluginOptions: pluginOptions || {},
      metadata,
      options: {}
    })
    debug(`Apply ${name}`)
    buffer = await plugin.apply(buffer)
  }

  return { content: buffer, metadata }
}

export default async function sqip(
  options: SqipConfig
): Promise<SqipResult | SqipResult[]> {
  // Build configuration based on passed options and default options
  const defaultOptions = {
    plugins: [
      { name: 'primitive', options: { numberOfPrimitives: 8, mode: 0 } },
      'blur',
      'svgo',
      'data-uri'
    ],
    width: 300,
    parseableOutput: false,
    silent: true
  }

  const config = Object.assign({}, defaultOptions, options)

  const { input, outputFileName, parseableOutput, silent } = config

  if (parseableOutput) {
    chalk.level = 0
  }

  // Validate configuration
  if (!input) {
    throw new Error(
      'Please provide an input image, e.g. sqip({ input: "input.jpg" })'
    )
  }

  // If input is a Buffer
  if (Buffer.isBuffer(input)) {
    if (!outputFileName) {
      throw new Error(
        `${outputFileName} is required when passing image as buffer`
      )
    }
    return processFile({
      buffer: input,
      outputFileName,
      config
    })
  }

  const files = await locateFiles(input)
  debug('Found files:')
  debug(files)

  // Test if all files are accessable
  for (const file of files) {
    try {
      debug('check file ' + file)
      await fs.access(file, fs.constants.R_OK)
    } catch (err) {
      throw new Error(`Unable to read file ${file}`)
    }
  }

  // Iterate over all files
  const results = []
  for (const filePath of files) {
    // Apply plugins to files
    if (!silent) {
      console.log(`Processing: ${filePath}`)
    } else {
      debug(`Processing ${filePath}`)
    }
    const buffer = await fs.readFile(filePath)
    const result = await processFile({
      buffer,
      outputFileName: outputFileName || path.parse(filePath).name,
      config
    })

    results.push(result)
  }

  debug(`Finished`)

  // Return as array when input was array or results is only one file
  if (Array.isArray(input) || results.length === 0) {
    return results
  }
  return results[0]
}

export * from './helpers'
