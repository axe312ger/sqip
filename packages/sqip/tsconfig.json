{
  "extends": "../../tsconfig.json",
  "compilerOptions": {
    "outDir": "dist",
    "rootDir": "src",
    "typeRoots": [
      "./typings",
      "./node_modules/@types",
      "../../node_modules/@types"
    ]
  },
  "exclude": ["node_modules", "dist", "__tests__"]
}
