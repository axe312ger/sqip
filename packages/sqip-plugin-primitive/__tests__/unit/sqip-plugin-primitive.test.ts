import execa, { ExecaChildProcess } from 'execa'
import fs from 'fs/promises'
import os from 'os'

import PrimitivePlugin from '../../src/sqip-plugin-primitive'
import { SqipImageMetadata, mockedMetadata } from 'sqip'

const sqipMockedMetadata: SqipImageMetadata = {
  ...mockedMetadata,
  type: 'pixel'
}

jest.mock('execa')
jest.mock('fs/promises')
jest.mock('os', () => ({
  ...(jest.requireActual('os') as typeof os),
  platform: jest.fn(() => 'unknownOS'),
  arch: jest.fn(() => 'nonExistingArch'),
  cpus: () => [1]
}))

const mockedExeca = execa as jest.MockedFunction<typeof execa>
mockedExeca.mockImplementation(() => {
  const result = {
    stdout:
      '<svg viewBox="0 0 1024 768"><rect fill="#bada5500"/><g></g></svg>'
  } as unknown as ExecaChildProcess<Buffer>
  return result
})

const mockedFsAccess = fs.access as jest.MockedFunction<typeof fs.access>

mockedFsAccess.mockImplementation(async () =>
  Promise.reject(new Error('Mocked: Binary not available'))
)
const mockedOsArch = os.arch as jest.MockedFunction<typeof os.arch>
const mockedOsPlatform = os.platform as jest.MockedFunction<typeof os.platform>

const proccessExitSpy = jest.spyOn(process, 'exit').mockImplementation()

const mockedConfig = {
  input: 'mocked',
  output: 'mocked',
  plugins: ['primitive']
}

describe('checkForPrimitive', () => {
  const primitivePlugin = new PrimitivePlugin({
    pluginOptions: {},
    options: {},
    sqipConfig: mockedConfig
  })

  afterEach(() => {
    mockedExeca.mockClear()
    mockedFsAccess.mockClear()
    mockedOsArch.mockClear()
    mockedOsPlatform.mockClear()
    proccessExitSpy.mockClear()
  })

  afterAll(() => {
    proccessExitSpy.mockReset()
  })

  test('bundled executable exists', async () => {
    mockedFsAccess.mockImplementationOnce(async () => Promise.resolve())

    await primitivePlugin.checkForPrimitive()

    expect(global.process.exit).not.toHaveBeenCalled()
    expect(mockedExeca).not.toHaveBeenCalled()
  })

  test('uses where for windows, type for POSIX', async () => {
    mockedOsPlatform.mockImplementationOnce(() => 'win32')
    await primitivePlugin.checkForPrimitive()
    expect(mockedExeca).toHaveBeenCalledWith('where', ['primitive'])
    expect(mockedFsAccess.mock.calls[0][0]).toMatch(/\.exe$/)

    mockedOsPlatform.mockImplementationOnce(() => 'linux')
    await primitivePlugin.checkForPrimitive()
    expect(mockedExeca).toHaveBeenCalledWith('type', ['primitive'])
  })

  test('bundled executable does not exist but primitive is globally installed', async () => {
    await expect(primitivePlugin.checkForPrimitive()).resolves.toBeUndefined()
  })

  test('bundled executable does not exist, primitive not installed globally', async () => {
    mockedExeca.mockImplementationOnce(() => {
      throw new Error('not installed')
    })

    await expect(
      primitivePlugin.checkForPrimitive()
    ).rejects.toThrowErrorMatchingSnapshot()
  })
})

describe('runPrimitive', () => {
  // A real PNG file, because the primitive plugin pipes the buffer through
  // sharp.
  //
  // Source: https://www.mjt.me.uk/posts/smallest-png/
  const fileContent = Buffer.from(
    'iVBORw0KGgoAAAANSUhEUgAAAQAAAAEAAQMAAABmvDolAAAAA1BMVEW10NBjBBbqAAAAH0lEQVRoge3BAQ0AAADCoPdPbQ43oAAAAAAAAAAAvg0hAAABmmDh1QAAAABJRU5ErkJggg==',
    'base64'
  )

  afterEach(() => {
    mockedExeca.mockClear()
  })

  test('executes primitive with default config', async () => {
    const primitivePlugin = new PrimitivePlugin({
      pluginOptions: {},
      options: {},
      sqipConfig: mockedConfig
    })
    await primitivePlugin.apply(fileContent, { ...sqipMockedMetadata })
    expect(mockedExeca.mock.calls).toHaveLength(2)
    expect(mockedExeca.mock.calls[1]).toHaveLength(3)
    expect(mockedExeca.mock.calls[1][1]).toMatchSnapshot()
  })

  test('executes primitive with custom config, applying default number of primitives', async () => {
    const primitivePlugin = new PrimitivePlugin({
      pluginOptions: { mode: 5 },
      options: {},

      sqipConfig: mockedConfig
    })
    await primitivePlugin.apply(fileContent, { ...sqipMockedMetadata })
    expect(mockedExeca.mock.calls).toHaveLength(2)
    expect(mockedExeca.mock.calls[1]).toHaveLength(3)
    expect(mockedExeca.mock.calls[1][1]).toMatchSnapshot()
  })

  test('executes primitive with landscape dimensions', async () => {
    const primitivePlugin = new PrimitivePlugin({
      pluginOptions: {},
      options: {},
      sqipConfig: mockedConfig
    })
    await primitivePlugin.apply(fileContent, {
      ...sqipMockedMetadata,
      width: 600,
      height: 300
    })
    expect(mockedExeca.mock.calls).toHaveLength(2)
    expect(mockedExeca.mock.calls[1]).toHaveLength(3)
    expect(mockedExeca.mock.calls[1][1]).toMatchSnapshot()
  })

  test('allows avg as value for background', async () => {
    const primitivePlugin = new PrimitivePlugin({
      pluginOptions: {
        background: 'avg'
      },
      options: {},
      sqipConfig: mockedConfig
    })
    await primitivePlugin.apply(fileContent, { ...sqipMockedMetadata })
    expect(mockedExeca.mock.calls).toHaveLength(2)
    expect(mockedExeca.mock.calls[1]).toHaveLength(3)
    expect(mockedExeca.mock.calls[1][1]).toMatchSnapshot()
  })

  test('allows hex as value for background', async () => {
    const primitivePlugin = new PrimitivePlugin({
      pluginOptions: {
        background: '#654321'
      },
      options: {},
      sqipConfig: mockedConfig
    })
    await primitivePlugin.apply(fileContent, { ...sqipMockedMetadata })
    expect(mockedExeca.mock.calls).toHaveLength(2)
    expect(mockedExeca.mock.calls[1]).toHaveLength(3)
    expect(mockedExeca.mock.calls[1][1]).toMatchSnapshot()
  })

  test('removes background rectangle when using fully transparent background', async () => {
    const primitivePlugin = new PrimitivePlugin({
      pluginOptions: {
        background: '#bada5500'
      },
      options: {},
      sqipConfig: mockedConfig
    })
    const res = await primitivePlugin.apply(fileContent, {
      ...sqipMockedMetadata
    })
    expect(res.toString()).toMatchSnapshot()
  })
})
