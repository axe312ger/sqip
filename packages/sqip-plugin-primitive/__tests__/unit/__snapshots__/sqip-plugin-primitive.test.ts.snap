// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`checkForPrimitive bundled executable does not exist, primitive not installed globally 1`] = `"Please ensure that Primitive (https://github.com/hashbite/primitive, written in Golang) is installed and globally available."`;

exports[`runPrimitive allows avg as value for background 1`] = `
Array [
  "-i",
  "-",
  "-o",
  "-",
  "-n",
  "8",
  "-m",
  "0",
  "-s",
  "1024",
  "-rep",
  "0",
  "-a",
  "128",
  "-bg",
  "avg",
  "-j",
  "0",
]
`;

exports[`runPrimitive allows hex as value for background 1`] = `
Array [
  "-i",
  "-",
  "-o",
  "-",
  "-n",
  "8",
  "-m",
  "0",
  "-s",
  "1024",
  "-rep",
  "0",
  "-a",
  "128",
  "-bg",
  "654321",
  "-j",
  "0",
]
`;

exports[`runPrimitive executes primitive with custom config, applying default number of primitives 1`] = `
Array [
  "-i",
  "-",
  "-o",
  "-",
  "-n",
  "8",
  "-m",
  "5",
  "-s",
  "1024",
  "-rep",
  "0",
  "-a",
  "128",
  "-bg",
  "040200",
  "-j",
  "0",
]
`;

exports[`runPrimitive executes primitive with default config 1`] = `
Array [
  "-i",
  "-",
  "-o",
  "-",
  "-n",
  "8",
  "-m",
  "0",
  "-s",
  "1024",
  "-rep",
  "0",
  "-a",
  "128",
  "-bg",
  "040200",
  "-j",
  "0",
]
`;

exports[`runPrimitive executes primitive with landscape dimensions 1`] = `
Array [
  "-i",
  "-",
  "-o",
  "-",
  "-n",
  "8",
  "-m",
  "0",
  "-s",
  "600",
  "-rep",
  "0",
  "-a",
  "128",
  "-bg",
  "040200",
  "-j",
  "0",
]
`;

exports[`runPrimitive removes background rectangle when using fully transparent background 1`] = `"<svg viewBox=\\"0 0 1024 768\\"><g></g></svg>"`;
